diesel/src/macros/static_cond.rs

```diff
@@ -21,18 +21,18 @@ macro_rules! static_cond {
 
```

Comment:
agreed, thanks for that catch

Reply:
@weiznich I had a look at doing this but couldn't get anywhere meaningful.
This is where I currently stashed it.

```diff
diff --git a/diesel/Cargo.toml b/diesel/Cargo.toml
index f5b79b05..440d9cae 100644
--- a/diesel/Cargo.toml
+++ b/diesel/Cargo.toml
@@ -14,6 +14,7 @@ categories = ["database"]
 [dependencies]
 byteorder = "1.0"
 diesel_derives = "~1.4.0"
+static-cond = "0.2.1"
 chrono = { version = "0.4", optional = true }
 libc = { version = "0.2.0", optional = true }
 libsqlite3-sys = { version = ">=0.8.0, <0.16.0", optional = true, features = ["min_sqlite_version_3_7_16"] }
diff --git a/diesel/src/lib.rs b/diesel/src/lib.rs
index 7dde8260..34c52b6a 100644
--- a/diesel/src/lib.rs
+++ b/diesel/src/lib.rs
@@ -141,6 +141,10 @@ pub use diesel_derives::{
 #[macro_use]
 mod macros;
 
+extern crate static_cond as sc;
+#[doc(inline)]
+pub use sc::static_cond as static_cond;
+

 #[cfg(test)]
 #[macro_use]
 extern crate cfg_if;
diff --git a/diesel/src/macros/mod.rs b/diesel/src/macros/mod.rs
index 8c608dc4..3365b7b8 100644
--- a/diesel/src/macros/mod.rs
+++ b/diesel/src/macros/mod.rs
@@ -609,6 +609,7 @@ macro_rules! __diesel_table_impl {
             use $crate::query_builder::nodes::Identifier;
             use $crate::query_source::{AppearsInFromClause, Once, Never};
             use $crate::query_source::joins::{Join, JoinOn};
+            use $crate::static_cond;
             $($imports)*
             pub use self::columns::*;
 
@@ -1060,8 +1061,6 @@ macro_rules! not_none {
 #[macro_use]
 mod internal;
 #[macro_use]
-mod static_cond;
-#[macro_use]
 mod ops;
 #[macro_use]
 mod tuples;
diff --git a/diesel/src/macros/static_cond.rs b/diesel/src/macros/static_cond.rs
deleted file mode 100644
index d59c78ee..00000000
--- a/diesel/src/macros/static_cond.rs
+++ /dev/null
```

We should also be able to pull this into a separate PR re-exporting has been supported since rust 1.30.0

I also tried something like this

```rust
pub mod macros {
    #[doc(inline)]
    pub use $super::macros::dothisforeachone;

    #[doc(inline)]
    pub use sc::static_cond as static_cond;
}
```

but the issue here is that we need macros to be defined above in order to access `$super::macros`

```
 #[macro_use]
 mod macros;
```

but then I am not allowed to redeclare `pub mod macros`

I think I can rename the the folder to something like `diesel_macros` and then pull that and re-export it as `pub mod macros {` pulling in each one.

Might not be the best solution and others will probably have more insight into what is best to do here.

I might also be missing a few bits here as the error messages are not the most helpful pushing me onto nightly to see more with `RUSTFLAGS="-Z external-macro-backtrace"` which didn't compile at all through cargo.

Reply:
I've played a bit around with using the `static_cond` crate here. It seems like that is not possible because the `static_cond` crate is incompatible with our current macro. In detail it does not allow to expand to an item while our own macro allows that. This is caused by a missing semicolon [there](](https://github.com/durka/static-cond/blob/master/src/lib.rs#L41)).

Reply:
```
diff --git a/diesel/Cargo.toml b/diesel/Cargo.toml
index f5b79b059..440d9cae4 100644
--- a/diesel/Cargo.toml
+++ b/diesel/Cargo.toml
@@ -14,6 +14,7 @@ categories = ["database"]
 [dependencies]
 byteorder = "1.0"
 diesel_derives = "~1.4.0"
+static-cond = "0.2.1"
 chrono = { version = "0.4", optional = true }
 libc = { version = "0.2.0", optional = true }
 libsqlite3-sys = { version = ">=0.8.0, <0.16.0", optional = true, features = ["min_sqlite_version_3_7_16"] }

diff --git a/diesel/src/lib.rs b/diesel/src/lib.rs
index e1f490dc4..8d30d334a 100644
--- a/diesel/src/lib.rs
+++ b/diesel/src/lib.rs
@@ -139,6 +139,9 @@ pub mod macros;
 #[macro_use]
 extern crate cfg_if;

+#[macro_use]
+extern crate static_cond;
+
 #[cfg(test)]
 pub mod test_helpers;


diff --git a/diesel/src/macros/mod.rs b/diesel/src/macros/mod.rs
index 5d86ba7cc..aabb8f920 100644
--- a/diesel/src/macros/mod.rs
+++ b/diesel/src/macros/mod.rs
@@ -616,12 +616,12 @@ macro_rules! __diesel_table_impl {
             use $crate::query_source::joins::{Join, JoinOn};
             $($imports)*
             pub use self::columns::*;
-
+            pub use static_cond::static_cond;
             /// Re-exports all of the columns of this table, as well as the
             /// table struct renamed to the module name. This is meant to be
             /// glob imported for functions which only deal with one table.
             pub mod dsl {
-                $($crate::static_cond! {
+                $(static_cond! {
                     if $table_name == $column_name {
                         compile_error!(concat!(
                             "Column `",
@@ -637,7 +637,7 @@ macro_rules! __diesel_table_impl {
                     } else {
                         pub use super::columns::{$column_name};
                     }
-                })+
+                };)+
                 pub use super::table as $table_name;
             }

@@ -1064,8 +1064,8 @@ macro_rules! not_none {
 // Utility macros which don't call any others need to come first.
 #[macro_use]
 mod internal;
-#[macro_use]
-mod static_cond;
+#[doc(inline)]
+pub use static_cond::static_cond;
 #[macro_use]
 mod ops;
 #[macro_use]
```

Agreed, I ended up with something like this in the end trying every combination to get around it.

lets resolve this as not possible for now